File: apollo-compiler/src/test/graphql/com/example/no_accessors/fragment/HeroDetails.java
package com.example.no_accessors.fragment;

import com.apollographql.apollo.api.Field;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.internal.Optional;
import java.io.IOException;
import java.lang.Integer;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import javax.annotation.Generated;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Generated("Apollo GraphQL")
public class HeroDetails {
  public static final String FRAGMENT_DEFINITION = "fragment HeroDetails on Character {\n"
      + "  __typename\n"
      + "  name\n"
      + "  friendsConnection {\n"
      + "    __typename\n"
      + "    totalCount\n"
      + "    edges {\n"
      + "      __typename\n"
      + "      node {\n"
      + "        __typename\n"
      + "        name\n"
      + "      }\n"
      + "    }\n"
      + "  }\n"
      + "  ... on Droid {\n"
      + "    __typename\n"
      + "    name\n"
      + "    primaryFunction\n"
      + "  }\n"
      + "}";

  public static final List<String> POSSIBLE_TYPES = Collections.unmodifiableList(Arrays.asList( "Human", "Droid"));

  public final Optional<AsDroid> asDroid;

  public HeroDetails() {
    if (this instanceof AsDroid) {
      asDroid = Optional.fromNullable((AsDroid) this);
    } else {
      asDroid = Optional.absent();
    }
  }

  @Override
  public String toString() {
    return "HeroDetails{"
      + "asDroid=" + asDroid
      + "}";
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (o instanceof HeroDetails) {
      HeroDetails that = (HeroDetails) o;
      return ((this.asDroid == null) ? (that.asDroid == null) : this.asDroid.equals(that.asDroid));
    }
    return false;
  }

  @Override
  public int hashCode() {
    int h = 1;
    h *= 1000003;
    h ^= (asDroid == null) ? 0 : asDroid.hashCode();
    return h;
  }

  public static final class Mapper implements ResponseFieldMapper<HeroDetails> {
    final AsDroid.Mapper asDroidFieldMapper = new AsDroid.Mapper();

    final Field[] fields = {
      Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<AsDroid>() {
        @Override
        public AsDroid read(String conditionalType, ResponseReader reader) throws IOException {
          if (conditionalType.equals("Droid")) {
            return asDroidFieldMapper.map(reader);
          } else {
            return null;
          }
        }
      })
    };

    @Override
    public HeroDetails map(ResponseReader reader) throws IOException {
      final AsDroid asDroid = reader.read(fields[0]);
      if (asDroid != null) {
        return asDroid;
      }
      return new HeroDetails();
    }
  }

  public static class Node {
    public final @Nonnull String name;

    public Node(@Nonnull String name) {
      this.name = name;
    }

    @Override
    public String toString() {
      return "Node{"
        + "name=" + name
        + "}";
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Node) {
        Node that = (Node) o;
        return ((this.name == null) ? (that.name == null) : this.name.equals(that.name));
      }
      return false;
    }

    @Override
    public int hashCode() {
      int h = 1;
      h *= 1000003;
      h ^= (name == null) ? 0 : name.hashCode();
      return h;
    }

    public static final class Mapper implements ResponseFieldMapper<Node> {
      final Field[] fields = {
        Field.forString("name", "name", null, false)
      };

      @Override
      public Node map(ResponseReader reader) throws IOException {
        final String name = reader.read(fields[0]);
        return new Node(name);
      }
    }
  }

  public static class Edge {
    public final Optional<Node> node;

    public Edge(@Nullable Node node) {
      this.node = Optional.fromNullable(node);
    }

    @Override
    public String toString() {
      return "Edge{"
        + "node=" + node
        + "}";
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Edge) {
        Edge that = (Edge) o;
        return ((this.node == null) ? (that.node == null) : this.node.equals(that.node));
      }
      return false;
    }

    @Override
    public int hashCode() {
      int h = 1;
      h *= 1000003;
      h ^= (node == null) ? 0 : node.hashCode();
      return h;
    }

    public static final class Mapper implements ResponseFieldMapper<Edge> {
      final Node.Mapper nodeFieldMapper = new Node.Mapper();

      final Field[] fields = {
        Field.forObject("node", "node", null, true, new Field.ObjectReader<Node>() {
          @Override public Node read(final ResponseReader reader) throws IOException {
            return nodeFieldMapper.map(reader);
          }
        })
      };

      @Override
      public Edge map(ResponseReader reader) throws IOException {
        final Node node = reader.read(fields[0]);
        return new Edge(node);
      }
    }
  }

  public static class FriendsConnection {
    public final Optional<Integer> totalCount;

    public final Optional<List<Edge>> edges;

    public FriendsConnection(@Nullable Integer totalCount, @Nullable List<Edge> edges) {
      this.totalCount = Optional.fromNullable(totalCount);
      this.edges = Optional.fromNullable(edges);
    }

    @Override
    public String toString() {
      return "FriendsConnection{"
        + "totalCount=" + totalCount + ", "
        + "edges=" + edges
        + "}";
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof FriendsConnection) {
        FriendsConnection that = (FriendsConnection) o;
        return ((this.totalCount == null) ? (that.totalCount == null) : this.totalCount.equals(that.totalCount))
         && ((this.edges == null) ? (that.edges == null) : this.edges.equals(that.edges));
      }
      return false;
    }

    @Override
    public int hashCode() {
      int h = 1;
      h *= 1000003;
      h ^= (totalCount == null) ? 0 : totalCount.hashCode();
      h *= 1000003;
      h ^= (edges == null) ? 0 : edges.hashCode();
      return h;
    }

    public static final class Mapper implements ResponseFieldMapper<FriendsConnection> {
      final Edge.Mapper edgeFieldMapper = new Edge.Mapper();

      final Field[] fields = {
        Field.forInt("totalCount", "totalCount", null, true),
        Field.forList("edges", "edges", null, true, new Field.ObjectReader<Edge>() {
          @Override public Edge read(final ResponseReader reader) throws IOException {
            return edgeFieldMapper.map(reader);
          }
        })
      };

      @Override
      public FriendsConnection map(ResponseReader reader) throws IOException {
        final Integer totalCount = reader.read(fields[0]);
        final List<Edge> edges = reader.read(fields[1]);
        return new FriendsConnection(totalCount, edges);
      }
    }
  }

  public static class AsDroid extends HeroDetails {
    public final @Nonnull String name;

    public final @Nonnull FriendsConnection friendsConnection;

    public final Optional<String> primaryFunction;

    public AsDroid(@Nonnull String name, @Nonnull FriendsConnection friendsConnection,
        @Nullable String primaryFunction) {
      this.name = name;
      this.friendsConnection = friendsConnection;
      this.primaryFunction = Optional.fromNullable(primaryFunction);
    }

    @Override
    public String toString() {
      return "AsDroid{"
        + "name=" + name + ", "
        + "friendsConnection=" + friendsConnection + ", "
        + "primaryFunction=" + primaryFunction
        + "}";
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof AsDroid) {
        AsDroid that = (AsDroid) o;
        return ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.friendsConnection == null) ? (that.friendsConnection == null) : this.friendsConnection.equals(that.friendsConnection))
         && ((this.primaryFunction == null) ? (that.primaryFunction == null) : this.primaryFunction.equals(that.primaryFunction));
      }
      return false;
    }

    @Override
    public int hashCode() {
      int h = 1;
      h *= 1000003;
      h ^= (name == null) ? 0 : name.hashCode();
      h *= 1000003;
      h ^= (friendsConnection == null) ? 0 : friendsConnection.hashCode();
      h *= 1000003;
      h ^= (primaryFunction == null) ? 0 : primaryFunction.hashCode();
      return h;
    }

    public static final class Mapper implements ResponseFieldMapper<AsDroid> {
      final FriendsConnection.Mapper friendsConnectionFieldMapper = new FriendsConnection.Mapper();

      final Field[] fields = {
        Field.forString("name", "name", null, false),
        Field.forObject("friendsConnection", "friendsConnection", null, false, new Field.ObjectReader<FriendsConnection>() {
          @Override public FriendsConnection read(final ResponseReader reader) throws IOException {
            return friendsConnectionFieldMapper.map(reader);
          }
        }),
        Field.forString("primaryFunction", "primaryFunction", null, true)
      };

      @Override
      public AsDroid map(ResponseReader reader) throws IOException {
        final String name = reader.read(fields[0]);
        final FriendsConnection friendsConnection = reader.read(fields[1]);
        final String primaryFunction = reader.read(fields[2]);
        return new AsDroid(name, friendsConnection, primaryFunction);
      }
    }
  }
}

File: apollo-compiler/src/test/graphql/com/example/java_beans_semantic_naming/fragment/HeroDetails.kt
package com.example.java_beans_semantic_naming.fragment

import com.apollographql.apollo.api.GraphqlFragment
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.ResponseFieldMarshaller
import com.apollographql.apollo.api.ResponseReader
import javax.annotation.Generated
import kotlin.Array
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List

/**
 * @param name The name of the character@param friendsConnection The friends of the character exposed as a connection with edges */
@Generated("Apollo GraphQL")
data class HeroDetails(
    val __typename: String,
    val name: String,
    val friendsConnection: FriendsConnection1,
    val asDroid: AsDroid?
) : GraphqlFragment {
    override fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
        it.writeString(RESPONSE_FIELDS[0], __typename)
        it.writeString(RESPONSE_FIELDS[1], name)
        it.writeObject(RESPONSE_FIELDS[2], friendsConnection.marshaller())
        it.writeObject(RESPONSE_FIELDS[3], asDroid?.marshaller())
    }

    companion object {
        private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                    ResponseField.forString("__typename", "__typename", null, false, null),
                    ResponseField.forString("name", "name", null, false, null),
                    ResponseField.forObject("friendsConnection", "friendsConnection", null, false, null),
                    ResponseField.forInlineFragment("__typename", "__typename", listOf("Droid"))
                )

        val FRAGMENT_DEFINITION: String = """
                |fragment HeroDetails on Character {
                |  __typename
                |  name
                |  friendsConnection {
                |    __typename
                |    totalCount
                |    edges {
                |      __typename
                |      node {
                |        __typename
                |        name
                |      }
                |    }
                |    pageInfo {
                |      __typename
                |      hasNextPage
                |    }
                |    isEmpty
                |  }
                |  ... on Droid {
                |    name
                |    primaryFunction
                |  }
                |}
                """.trimMargin()

        val POSSIBLE_TYPES: Array<String> = arrayOf("Human", "Droid")

        operator fun invoke(reader: ResponseReader): HeroDetails {
            val __typename = reader.readString(RESPONSE_FIELDS[0])
            val name = reader.readString(RESPONSE_FIELDS[1])
            val friendsConnection = reader.readObject<FriendsConnection1>(RESPONSE_FIELDS[2]) {
                FriendsConnection1(it)
            }

            val asDroid = reader.readConditional(RESPONSE_FIELDS[3]) { conditionalType, reader ->
                AsDroid(reader)
            }

            return HeroDetails(
                __typename = __typename,
                name = name,
                friendsConnection = friendsConnection,
                asDroid = asDroid
            )
        }
    }

    /**
     * @param name The name of the character */
    data class Node(
        val __typename: String,
        val name: String
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeString(RESPONSE_FIELDS[1], name)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forString("name", "name", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): Node {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val name = reader.readString(RESPONSE_FIELDS[1])
                return Node(
                    __typename = __typename,
                    name = name
                )
            }
        }
    }

    /**
     * @param node The character represented by this friendship edge */
    data class Edges(
        val __typename: String,
        val node: Node?
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeObject(RESPONSE_FIELDS[1], node?.marshaller())
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forObject("node", "node", null, true, null)
                    )

            operator fun invoke(reader: ResponseReader): Edges {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val node = reader.readObject<Node>(RESPONSE_FIELDS[1]) {
                    Node(it)
                }

                return Edges(
                    __typename = __typename,
                    node = node
                )
            }
        }
    }

    data class PageInfo(val __typename: String, val hasNextPage: Boolean) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeBoolean(RESPONSE_FIELDS[1], hasNextPage)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forBoolean("hasNextPage", "hasNextPage", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): PageInfo {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val hasNextPage = reader.readBoolean(RESPONSE_FIELDS[1])
                return PageInfo(
                    __typename = __typename,
                    hasNextPage = hasNextPage
                )
            }
        }
    }

    /**
     * @param totalCount The total number of friends@param edges The edges for each of the character's friends.@param pageInfo Information for paginating this connection@param isEmpty For test java beans semantic naming only */
    data class FriendsConnection(
        val __typename: String,
        val totalCount: Int?,
        val edges: List<Edges>?,
        val pageInfo: PageInfo,
        val isEmpty: Boolean
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeInt(RESPONSE_FIELDS[1], totalCount)
            it.writeList(RESPONSE_FIELDS[2], edges) { value, listItemWriter ->
                        @Suppress("NAME_SHADOWING")
                        value?.forEach { value ->
                            listItemWriter.writeObject(value?.marshaller())}
                    }

            it.writeObject(RESPONSE_FIELDS[3], pageInfo.marshaller())
            it.writeBoolean(RESPONSE_FIELDS[4], isEmpty)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forInt("totalCount", "totalCount", null, true, null),
                        ResponseField.forList("edges", "edges", null, true, null),
                        ResponseField.forObject("pageInfo", "pageInfo", null, false, null),
                        ResponseField.forBoolean("isEmpty", "isEmpty", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): FriendsConnection {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val totalCount = reader.readInt(RESPONSE_FIELDS[1])
                val edges = reader.readList<Edges>(RESPONSE_FIELDS[2]) {
                    it.readObject<Edges> {
                        Edges(it)
                    }

                }

                val pageInfo = reader.readObject<PageInfo>(RESPONSE_FIELDS[3]) {
                    PageInfo(it)
                }

                val isEmpty = reader.readBoolean(RESPONSE_FIELDS[4])
                return FriendsConnection(
                    __typename = __typename,
                    totalCount = totalCount,
                    edges = edges,
                    pageInfo = pageInfo,
                    isEmpty = isEmpty
                )
            }
        }
    }

    /**
     * @param name What others call this droid@param friendsConnection The friends of the droid exposed as a connection with edges@param primaryFunction This droid's primary function */
    data class AsDroid(
        val __typename: String,
        val name: String,
        val friendsConnection: FriendsConnection,
        val primaryFunction: String?
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeString(RESPONSE_FIELDS[1], name)
            it.writeObject(RESPONSE_FIELDS[2], friendsConnection.marshaller())
            it.writeString(RESPONSE_FIELDS[3], primaryFunction)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forString("name", "name", null, false, null),
                        ResponseField.forObject("friendsConnection", "friendsConnection", null, false, null),
                        ResponseField.forString("primaryFunction", "primaryFunction", null, true, null)
                    )

            operator fun invoke(reader: ResponseReader): AsDroid {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val name = reader.readString(RESPONSE_FIELDS[1])
                val friendsConnection = reader.readObject<FriendsConnection>(RESPONSE_FIELDS[2]) {
                    FriendsConnection(it)
                }

                val primaryFunction = reader.readString(RESPONSE_FIELDS[3])
                return AsDroid(
                    __typename = __typename,
                    name = name,
                    friendsConnection = friendsConnection,
                    primaryFunction = primaryFunction
                )
            }
        }
    }

    /**
     * @param name The name of the character */
    data class Node1(
        val __typename: String,
        val name: String
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeString(RESPONSE_FIELDS[1], name)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forString("name", "name", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): Node1 {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val name = reader.readString(RESPONSE_FIELDS[1])
                return Node1(
                    __typename = __typename,
                    name = name
                )
            }
        }
    }

    /**
     * @param node The character represented by this friendship edge */
    data class Edges1(
        val __typename: String,
        val node: Node1?
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeObject(RESPONSE_FIELDS[1], node?.marshaller())
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forObject("node", "node", null, true, null)
                    )

            operator fun invoke(reader: ResponseReader): Edges1 {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val node = reader.readObject<Node1>(RESPONSE_FIELDS[1]) {
                    Node1(it)
                }

                return Edges1(
                    __typename = __typename,
                    node = node
                )
            }
        }
    }

    data class PageInfo1(val __typename: String, val hasNextPage: Boolean) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeBoolean(RESPONSE_FIELDS[1], hasNextPage)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forBoolean("hasNextPage", "hasNextPage", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): PageInfo1 {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val hasNextPage = reader.readBoolean(RESPONSE_FIELDS[1])
                return PageInfo1(
                    __typename = __typename,
                    hasNextPage = hasNextPage
                )
            }
        }
    }

    /**
     * @param totalCount The total number of friends@param edges The edges for each of the character's friends.@param pageInfo Information for paginating this connection@param isEmpty For test java beans semantic naming only */
    data class FriendsConnection1(
        val __typename: String,
        val totalCount: Int?,
        val edges: List<Edges1>?,
        val pageInfo: PageInfo1,
        val isEmpty: Boolean
    ) {
        fun marshaller(): ResponseFieldMarshaller = ResponseFieldMarshaller {
            it.writeString(RESPONSE_FIELDS[0], __typename)
            it.writeInt(RESPONSE_FIELDS[1], totalCount)
            it.writeList(RESPONSE_FIELDS[2], edges) { value, listItemWriter ->
                        @Suppress("NAME_SHADOWING")
                        value?.forEach { value ->
                            listItemWriter.writeObject(value?.marshaller())}
                    }

            it.writeObject(RESPONSE_FIELDS[3], pageInfo.marshaller())
            it.writeBoolean(RESPONSE_FIELDS[4], isEmpty)
        }

        companion object {
            private val RESPONSE_FIELDS: Array<ResponseField> = arrayOf(
                        ResponseField.forString("__typename", "__typename", null, false, null),
                        ResponseField.forInt("totalCount", "totalCount", null, true, null),
                        ResponseField.forList("edges", "edges", null, true, null),
                        ResponseField.forObject("pageInfo", "pageInfo", null, false, null),
                        ResponseField.forBoolean("isEmpty", "isEmpty", null, false, null)
                    )

            operator fun invoke(reader: ResponseReader): FriendsConnection1 {
                val __typename = reader.readString(RESPONSE_FIELDS[0])
                val totalCount = reader.readInt(RESPONSE_FIELDS[1])
                val edges = reader.readList<Edges1>(RESPONSE_FIELDS[2]) {
                    it.readObject<Edges1> {
                        Edges1(it)
                    }

                }

                val pageInfo = reader.readObject<PageInfo1>(RESPONSE_FIELDS[3]) {
                    PageInfo1(it)
                }

                val isEmpty = reader.readBoolean(RESPONSE_FIELDS[4])
                return FriendsConnection1(
                    __typename = __typename,
                    totalCount = totalCount,
                    edges = edges,
                    pageInfo = pageInfo,
                    isEmpty = isEmpty
                )
            }
        }
    }
}
